ledger-core/inc/core/math/bech32/Bech32ParametersHelper.hpp

```diff
@@ -0,0 +1,60 @@
+/*
+ *
+ * Bech32ParametersHelper
+ *
+ * Created by Gerry Agbobada on 23/01/2020
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2020 Ledger
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+#pragma once
+
+#ifndef LIBCORE_EXPORT
+    #if defined(_MSC_VER)
+        #include <libcore_export.h>
+    #else
+        #define LIBCORE_EXPORT
+    #endif
+#endif
+
+#include <core/math/bech32/Bech32Parameters.hpp>
+
+namespace ledger {
+    namespace core {
+        // CRTP used to get static polymorphism on the interface
+        template <class CoinLikeBech32ParametersHelper>
+        class Bech32ParametersHelper {
+            public:
+                static const Bech32Parameters::Bech32Struct getBech32Params(const std::string &networkIdentifier) {
+                    return CoinLikeBech32ParametersHelper::getCoinLikeBech32Params(networkIdentifier);
```

Comment:
Thats method is just a forward that doesn’t add any value. I struggle a bit to understand the goal of that CRTP class. Can you explain a bit why you need it and what it tries to solve? Thanks! 🙂

Reply:
Bech32 interface (the virtual methods [here](https://github.com/LedgerHQ/lib-ledger-core/blob/1bd439d5d9b05025259f788f0ff9d99c8571821f/ledger-core/inc/core/math/bech32/Bech32.hpp#L55-L74) ) is implemented by coins using coin-specific parameters ( defined [here](https://github.com/LedgerHQ/lib-ledger-core/blob/1bd439d5d9b05025259f788f0ff9d99c8571821f/ledger-core-bitcoin/src/bitcoin/bech32/BitcoinLikeBech32ParametersHelper.cpp#L44) for bitcoin )

The implementors of Bech32 interface (BTCBech32 example below, it's the same of BCHBech32) are using a call to a static `getBech32Params` method to get the correct private parameters attribute.
https://github.com/LedgerHQ/lib-ledger-core/blob/1bd439d5d9b05025259f788f0ff9d99c8571821f/ledger-core-bitcoin/inc/bitcoin/bech32/BTCBech32.hpp#L42

(I also noticed that I forgot another usage in `BitcoinLikeAddress` for [getVersionFromKeychainEngine code](https://github.com/LedgerHQ/lib-ledger-core/blob/1bd439d5d9b05025259f788f0ff9d99c8571821f/ledger-core-bitcoin/src/bitcoin/BitcoinLikeAddress.cpp#L77), I'm having issues compiling DerivationPath recently so I wasn't able to properly produce working tests for now :( )

From this point :
- All usages of `Bech32Parameters` come from static methods getBech32Params(string&) and insertParameters(..) ; and static member `ALL`
- Those static members depend on static `Bech32Parameters::Bech32Struct` which should stay coin-specific and only instantiated in coin-specific libraries (bitcoin specific params are defined [here](https://github.com/LedgerHQ/lib-ledger-core/blob/1bd439d5d9b05025259f788f0ff9d99c8571821f/ledger-core-bitcoin/src/bitcoin/bech32/BitcoinLikeBech32ParametersHelper.cpp#L44) - for example cosmos specific ones look like [this](https://github.com/gagbo/lib-ledger-core/blob/9820248bd409e2b8039f16c999d43af30abf731a/ledger-core-cosmos/inc/cosmos/bech32/CosmosBech32.hpp#L48) )

So the goal with CRTP was to provide a way to specify a `static getBech32Params(string&)` interface to use in coins.

Using classic inheritance here means that client code has to change to provide an instance of `BitcoinLikeBech32Parameters` to override the non-static `getBech32Params(string&)` method